### PR TITLE
Update to modern javascript

### PR DESCRIPTION
**Pull request by AndyOGo**
- [x] Rewrite to ES6
- [x] add Babel 7
- [x] add rollup
- [x] update QUnit
- [x] add prettier, eslint, commit hook

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,25 +1,61 @@
 {
-  "name": "json-logic-js",
+  "name": "@axa-ch/json-logic-js",
   "version": "1.2.2",
```
Comment (AndyOGo) [inline package.json:3]: I would bump the version to `2.0.0` 🤔

**Comment by AndyOGo** [comment]
@mjenny 
@lina-salih 
I searched through the open PRs and found one which rewrote it to ES6 already, just not using modules 🤔 
https://github.com/jwadhams/json-logic-js/pull/56

File: lib/operations/accessor/missing.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (value === null || value === '') {
+      are_missing.push(key);
+    }
+  }
+
+  return are_missing;
+}
+
+missing.withApply = true;
```
Comment (AndyOGo) [inline lib/operations/accessor/missing.js:32]: @lina-salih 
@mjenny 
I don't like this hacks of meta data like `withApply` and `code`...
The reason I have them is, that some operations need to call `apply` recursively, and others have names within the API like `==` which is not a valid JavaScript function or variable name., so:
- `withApply`: adds the `apply` function as a callback as the first argument
- `code`: overrides the functions key within the `operations` object, e.g. `equal()` function gets overridden by it's `code` property being `==`

This is too complex and error prone 💣

File: lib/operations/accessor/missingSome.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  if (options.length - are_missing.length >= need_count) {
+    return [];
+  }
+
+  return are_missing;
+}
+
+missingSome.code = 'missing_some';
+missingSome.withApply = true;
```
Comment (AndyOGo) [inline lib/operations/accessor/missingSome.js:15]: @lina-salih @mjenny 
Same here `code`, `withApply` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

File: lib/operations/accessor/variable.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    if (data === undefined) {
+      return not_found;
+    }
+  }
+
+  return data;
+}
+
+variable.code = 'var';
```
Comment (AndyOGo) [inline lib/operations/accessor/variable.js:27]: @lina-salih @mjenny 
Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`var` is a reserved keyword and can't be used as function name

File: lib/operations/arithmetic/add.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
+    args[_key] = arguments[_key];
+  }
+
+  return args.reduce(function (a, b) {
+    return parseFloat(a, 10) + parseFloat(b, 10);
+  }, 0);
+}
+
+add.code = '+';
```
Comment (AndyOGo) [inline lib/operations/arithmetic/add.js:11]: @lina-salih @mjenny 
Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`+` is an existing operator

File: lib/operations/arithmetic/divide.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+function divide(a, b) {
+  return a / b;
+}
+
+divide.code = '/';
```
Comment (AndyOGo) [inline lib/operations/arithmetic/divide.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`/` is an existing operator

File: lib/operations/arithmetic/modulo.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+function modulo(a, b) {
+  return a % b;
+}
+
+modulo.code = '%';
```
Comment (AndyOGo) [inline lib/operations/arithmetic/modulo.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`%` is an existing operator

File: lib/operations/arithmetic/substract.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+function substract(a, b) {
+  if (b === undefined) {
+    return -a;
+  }
+
+  return a - b;
+}
+
+substract.code = '-';
```
Comment (AndyOGo) [inline lib/operations/arithmetic/substract.js:9]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`-` is an existing operator

File: lib/operations/logic/equal.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+function equal(a, b) {
+  // eslint-disable-next-line eqeqeq
+  return a == b;
+}
+
+equal.code = '==';
```
Comment (AndyOGo) [inline lib/operations/logic/equal.js:6]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`==` is an existing operator

File: lib/operations/logic/falsy.js [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import truthy from '../../helpers/truthy';
+
+function falsy(a) {
+  return !truthy(a);
+}
+
+falsy.code = '!';
```
Comment (AndyOGo) [inline lib/operations/logic/falsy.js:7]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`!` is an existing operator

File: lib/operations/logic/notEqual.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+function notEqual(a, b) {
+  // eslint-disable-next-line eqeqeq
+  return a != b;
+}
+
+notEqual.code = '!=';
```
Comment (AndyOGo) [inline lib/operations/logic/notEqual.js:6]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`!=` is an existing operator

File: lib/operations/logic/strictEqual.js [hunk 12]
```diff
@@ -0,0 +1,6 @@
+function strictEqual(a, b) {
+  return a === b;
+}
+
+strictEqual.code = '===';
```
Comment (AndyOGo) [inline lib/operations/logic/strictEqual.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`===` is an existing operator

File: lib/operations/logic/strictNotEqual.js [hunk 13]
```diff
@@ -0,0 +1,6 @@
+function strictNotEqual(a, b) {
+  return a !== b;
+}
+
+strictNotEqual.code = '!==';
```
Comment (AndyOGo) [inline lib/operations/logic/strictNotEqual.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`!==` is an existing operator

File: lib/operations/logic/truthy.js [hunk 14]
```diff
@@ -0,0 +1,3 @@
+import truthy from '../../helpers/truthy';
+truthy.code = '!!';
```
Comment (AndyOGo) [inline lib/operations/logic/truthy.js:2]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`!` is an existing operator

File: lib/operations/misc/indexOf.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
+function indexOf(a, b) {
+  if (!b || typeof b.indexOf === 'undefined') return false;
+  return b.indexOf(a) !== -1;
+}
+
+indexOf.code = 'in';
```
Comment (AndyOGo) [inline lib/operations/misc/indexOf.js:6]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`in` is an existing operator

File: lib/operations/numeric/greater.js [hunk 16]
```diff
@@ -0,0 +1,6 @@
+function greater(a, b) {
+  return a > b;
+}
+
+greater.code = '>';
```
Comment (AndyOGo) [inline lib/operations/numeric/greater.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`>` is an existing operator

File: lib/operations/numeric/greaterEqual.js [hunk 17]
```diff
@@ -0,0 +1,6 @@
+function greaterEqual(a, b) {
+  return a >= b;
+}
+
+greaterEqual.code = '>=';
```
Comment (AndyOGo) [inline lib/operations/numeric/greaterEqual.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`>=` is an existing operator

File: lib/operations/numeric/lower.js [hunk 18]
```diff
@@ -0,0 +1,6 @@
+function lower(a, b, c) {
+  return c === undefined ? a < b : a < b && b < c;
+}
+
+lower.code = '<';
```
Comment (AndyOGo) [inline lib/operations/numeric/lower.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`<` is an existing operator

File: lib/operations/numeric/lowerEqual.js [hunk 19]
```diff
@@ -0,0 +1,6 @@
+function lowerEqual(a, b, c) {
+  return c === undefined ? a <= b : a <= b && b <= c;
+}
+
+lowerEqual.code = '<=';
```
Comment (AndyOGo) [inline lib/operations/numeric/lowerEqual.js:5]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`<=` is an existing operator

File: lib/visitors/logic/condition.js [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  if (values.length === i + 1) {
+    return apply(values[i], data);
+  }
+
+  return null;
+}
+
+condition.code = ['if', '?:'];
```
Comment (AndyOGo) [inline lib/visitors/logic/condition.js:31]: Same here `code` as described in https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014

`/if` is an existing operator and `?:` not a valid variable name and it's and array...

File: lib/operations/accessor/missing.js (hunk 2 above)
Comment (storm1ng) [reply lib/operations/accessor/missing.js:32]: Do you plan on changing this at some point?

**Review by storm1ng — APPROVED** [review]
Other than https://github.com/axa-ch/json-logic-js/pull/1/files#r259730014 this LGTM

File: lib/operations/accessor/missing.js (hunk 2 above)
Comment (AndyOGo) [reply lib/operations/accessor/missing.js:32]: Yeah, it sit's on my nerves^^

**Comment by storm1ng** [comment]
Awesome PR, using once again up-to-date components 👍

File: lib/operations/accessor/missing.js (hunk 2 above)
Comment (storm1ng) [reply lib/operations/accessor/missing.js:32]: Go for it. 💯

Comment (l-salih) [reply lib/operations/accessor/missing.js:32]: Oh.... do you want to change this in the current PR or open an issue? 😖

File: lib/operations/arithmetic/add.js [hunk 21]
```diff
@@ -0,0 +1,12 @@
+function add() {
```
Comment (l-salih) [inline lib/operations/arithmetic/add.js:1]: do we need arithmetic operations?

File: lib/operations/logic/strictEqual.js (hunk 12 above)
Comment (l-salih) [reply lib/operations/logic/strictEqual.js:5]: yes 👍

File: src/createJsonLogic.js [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+  return {
+    apply,
+    add_operation: addOperation,
+    rm_operation: removeOperation,
+    add_visitor: addVisitor,
+    rm_visitor: removeVisitor,
+  };
+}
+
+export default createJsonLogic;
```
Comment (l-salih) [inline src/createJsonLogic.js:120]: awesome

File: src/operations/accessor/variable.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
+function variable(a, b) {
+  const not_found = b === undefined ? null : b;
+  let data = this;
+
+  if (typeof a === 'undefined' || a === '' || a === null) {
```
Comment (l-salih) [inline src/operations/accessor/variable.js:5]: could we export this as a function?

**Review by l-salih — APPROVED** [review]
Great job! thank you @AndyOGo 😄

File: lib/operations/arithmetic/add.js (hunk 21 above)
Comment (AndyOGo) [reply lib/operations/arithmetic/add.js:1]: I would kick them

File: src/operations/accessor/variable.js (hunk 23 above)
Comment (AndyOGo) [reply src/operations/accessor/variable.js:5]: sure as a helper

**Pull request merged**